File: lc_js/lcJs/design/lruCache.ts
// LeetCode 146: LRU缓存
// 设计一个数据结构，支持 get 和 put 操作，并且在容量满时淘汰最近最少使用的元素

class Node {
    key: number;
    val: number;
    prev: Node | null;
    next: Node | null;
    
    constructor(key: number = 0, val: number = 0) {
        this.key = key;
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

class LRUCache {
    private cap: number;
    private cache: Map<number, Node>;
    private head: Node;
    private tail: Node;
    
    constructor(capacity: number) {
        this.cap = capacity;
        this.cache = new Map<number, Node>();
        this.head = new Node();
        this.tail = new Node();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
    
    private addToHead(node: Node): void {
        node.prev = this.head;
        node.next = this.head.next;
        this.head.next!.prev = node;
        this.head.next = node;
    }
    
    private removeNode(node: Node): void {
        node.prev!.next = node.next;
        node.next!.prev = node.prev;
    }
    
    private moveToHead(node: Node): void {
        this.removeNode(node);
        this.addToHead(node);
    }
    
    private removeTail(): Node {
        const node = this.tail.prev!;
        this.removeNode(node);
        return node;
    }
    
    get(key: number): number {
        const node = this.cache.get(key);
        if (node) {
            this.moveToHead(node);
            return node.val;
        }
        return -1;
    }
    
    put(key: number, value: number): void {
        const node = this.cache.get(key);
        if (node) {
            node.val = value;
            this.moveToHead(node);
        } else {
            const newNode = new Node(key, value);
            this.cache.set(key, newNode);
            this.addToHead(newNode);
            if (this.cache.size > this.cap) {
                const removed = this.removeTail();
                this.cache.delete(removed.key);
            }
        }
    }
}

// 泛型版本的LRU缓存
class GenericLRUCache<K, V> {
    private capacity: number;
    private cache: Map<K, V>;
    
    constructor(capacity: number) {
        this.capacity = capacity;
        this.cache = new Map<K, V>();
    }
    
    get(key: K): V | undefined {
        if (this.cache.has(key)) {
            const value = this.cache.get(key)!;
            // 删除后重新插入以更新顺序
            this.cache.delete(key);
            this.cache.set(key, value);
            return value;
        }
        return undefined;
    }
    
    put(key: K, value: V): void {
        if (this.cache.has(key)) {
            this.cache.delete(key);
        } else if (this.cache.size >= this.capacity) {
            // 删除最旧的（第一个）
            const firstKey = this.cache.keys().next().value;
            this.cache.delete(firstKey);
        }
        this.cache.set(key, value);
    }
    
    has(key: K): boolean {
        return this.cache.has(key);
    }
    
    size(): number {
        return this.cache.size;
    }
    
    clear(): void {
        this.cache.clear();
    }
}

// 接口定义
interface ILRUCache<K, V> {
    get(key: K): V | undefined;
    put(key: K, value: V): void;
    has(key: K): boolean;
    size(): number;
    clear(): void;
}

// 实现接口的类
class TypedLRUCache<K, V> implements ILRUCache<K, V> {
    private capacity: number;
    private cache: Map<K, V>;
    
    constructor(capacity: number) {
        this.capacity = capacity;
        this.cache = new Map<K, V>();
    }
    
    get(key: K): V | undefined {
        if (this.cache.has(key)) {
            const value = this.cache.get(key)!;
            this.cache.delete(key);
            this.cache.set(key, value);
            return value;
        }
        return undefined;
    }
    
    put(key: K, value: V): void {
        if (this.cache.has(key)) {
            this.cache.delete(key);
        } else if (this.cache.size >= this.capacity) {
            const firstKey = this.cache.keys().next().value;
            this.cache.delete(firstKey);
        }
        this.cache.set(key, value);
    }
    
    has(key: K): boolean {
        return this.cache.has(key);
    }
    
    size(): number {
        return this.cache.size;
    }
    
    clear(): void {
        this.cache.clear();
    }
}

// 类型别名
type CacheEntry<T> = {
    key: string;
    value: T;
    timestamp: number;
};

// 带过期时间的LRU缓存
class TTLLRUCache<T> {
    private capacity: number;
    private ttl: number;
    private cache: Map<string, CacheEntry<T>>;
    
    constructor(capacity: number, ttl: number = 60000) {
        this.capacity = capacity;
        this.ttl = ttl;
        this.cache = new Map<string, CacheEntry<T>>();
    }
    
    get(key: string): T | undefined {
        const entry = this.cache.get(key);
        if (entry) {
            if (Date.now() - entry.timestamp > this.ttl) {
                this.cache.delete(key);
                return undefined;
            }
            // 更新访问时间并移动到最后
            entry.timestamp = Date.now();
            this.cache.delete(key);
            this.cache.set(key, entry);
            return entry.value;
        }
        return undefined;
    }
    
    put(key: string, value: T): void {
        const entry: CacheEntry<T> = {
            key,
            value,
            timestamp: Date.now()
        };
        
        if (this.cache.has(key)) {
            this.cache.delete(key);
        } else if (this.cache.size >= this.capacity) {
            const firstKey = this.cache.keys().next().value;
            this.cache.delete(firstKey);
        }
        
        this.cache.set(key, entry);
    }
}

function testLRUCache(): void {
    console.log("=== LRU缓存测试 ===");
    
    // 测试标准版本
    console.log("\n--- 标准版本测试 ---");
    const lru: LRUCache = new LRUCache(2);
    
    lru.put(1, 1);
    lru.put(2, 2);
    console.log(`get(1): ${lru.get(1)}`); // 返回 1
    lru.put(3, 3); // 该操作会使得关键字 2 作废
    console.log(`get(2): ${lru.get(2)}`); // 返回 -1 (未找到)
    lru.put(4, 4); // 该操作会使得关键字 1 作废
    console.log(`get(1): ${lru.get(1)}`); // 返回 -1 (未找到)
    console.log(`get(3): ${lru.get(3)}`); // 返回 3
    console.log(`get(4): ${lru.get(4)}`); // 返回 4
    
    // 测试泛型版本
    console.log("\n--- 泛型版本测试 ---");
    const genericLru: GenericLRUCache<string, number> = new GenericLRUCache<string, number>(2);
    
    genericLru.put("a", 1);
    genericLru.put("b", 2);
    console.log(`get("a"): ${genericLru.get("a")}`); // 返回 1
    genericLru.put("c", 3); // 该操作会使得关键字 "b" 作废
    console.log(`get("b"): ${genericLru.get("b")}`); // 返回 undefined
    
    // 测试TTL版本
    console.log("\n--- TTL版本测试 ---");
    const ttlLru: TTLLRUCache<string> = new TTLLRUCache<string>(2, 1000);
    
    ttlLru.put("key1", "value1");
    console.log(`get("key1"): ${ttlLru.get("key1")}`); // 返回 "value1"
    
    console.log("\n=== TypeScript特性说明 ===");
    console.log("1. 静态类型: 编译时类型检查");
    console.log("2. 泛型: 类型安全的代码复用");
    console.log("3. 接口: 契约定义");
    console.log("4. 类型别名: 复杂类型简化");
    console.log("5. 可选属性: 灵活的类型定义");
}

// 导出
export {
    LRUCache,
    GenericLRUCache,
    TypedLRUCache,
    TTLLRUCache,
    ILRUCache,
    CacheEntry,
    testLRUCache
};

// 默认导出
export default LRUCache;

// 如果在Node.js环境中直接运行
if (typeof require !== 'undefined' && require.main === module) {
    testLRUCache();
}

/*
TypeScript实现特点:

1. **静态类型系统**:
   - 编译时类型检查
   - 类型推断
   - 严格的类型安全

2. **面向对象特性**:
   - 类和接口
   - 继承和多态
   - 访问修饰符

3. **泛型编程**:
   - 类型参数
   - 约束泛型
   - 类型安全的代码复用

4. **高级类型**:
   - 联合类型
   - 交叉类型
   - 条件类型

5. **模块系统**:
   - ES6模块
   - 命名空间
   - 声明合并

算法复杂度:
- get操作: O(1)
- put操作: O(1)
- 空间复杂度: O(capacity)
*/

File: lc_cpp/lcCpp/design/lruCache.c
// LeetCode 146: LRU缓存
// 设计一个数据结构，支持 get 和 put 操作，并且在容量满时淘汰最近最少使用的元素

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define HASH_SIZE 10007

typedef struct Node {
    int key;
    int val;
    struct Node* prev;
    struct Node* next;
} Node;

typedef struct {
    int cap;
    int size;
    Node* head;
    Node* tail;
    Node* hash[HASH_SIZE];
} LRUCache;

Node* createNode(int key, int val) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->key = key;
    node->val = val;
    node->prev = NULL;
    node->next = NULL;
    return node;
}

int hash(int key) {
    return (key % HASH_SIZE + HASH_SIZE) % HASH_SIZE;
}

Node* findInHash(LRUCache* obj, int key) {
    int idx = hash(key);
    Node* curr = obj->hash[idx];
    while (curr) {
        if (curr->key == key) {
            return curr;
        }
        curr = curr->next;
    }
    return NULL;
}

void addToHash(LRUCache* obj, Node* node) {
    int idx = hash(node->key);
    node->next = obj->hash[idx];
    obj->hash[idx] = node;
}

void removeFromHash(LRUCache* obj, int key) {
    int idx = hash(key);
    Node* curr = obj->hash[idx];
    Node* prev = NULL;
    
    while (curr) {
        if (curr->key == key) {
            if (prev) {
                prev->next = curr->next;
            } else {
                obj->hash[idx] = curr->next;
            }
            break;
        }
        prev = curr;
        curr = curr->next;
    }
}

void addToHead(LRUCache* obj, Node* node) {
    node->prev = obj->head;
    node->next = obj->head->next;
    obj->head->next->prev = node;
    obj->head->next = node;
}

void removeNode(Node* node) {
    node->prev->next = node->next;
    node->next->prev = node->prev;
}

void moveToHead(LRUCache* obj, Node* node) {
    removeNode(node);
    addToHead(obj, node);
}

Node* removeTail(LRUCache* obj) {
    Node* node = obj->tail->prev;
    removeNode(node);
    return node;
}

LRUCache* lRUCacheCreate(int capacity) {
    LRUCache* obj = (LRUCache*)malloc(sizeof(LRUCache));
    obj->cap = capacity;
    obj->size = 0;
    obj->head = createNode(0, 0);
    obj->tail = createNode(0, 0);
    obj->head->next = obj->tail;
    obj->tail->prev = obj->head;
    
    memset(obj->hash, 0, sizeof(obj->hash));
    
    return obj;
}

int lRUCacheGet(LRUCache* obj, int key) {
    Node* node = findInHash(obj, key);
    if (node) {
        moveToHead(obj, node);
        return node->val;
    }
    return -1;
}

void lRUCachePut(LRUCache* obj, int key, int value) {
    Node* node = findInHash(obj, key);
    if (node) {
        node->val = value;
        moveToHead(obj, node);
    } else {
        Node* newNode = createNode(key, value);
        addToHash(obj, newNode);
        addToHead(obj, newNode);
        obj->size++;
        
        if (obj->size > obj->cap) {
            Node* removed = removeTail(obj);
            removeFromHash(obj, removed->key);
            free(removed);
            obj->size--;
        }
    }
}

void lRUCacheFree(LRUCache* obj) {
    Node* curr = obj->head;
    while (curr) {
        Node* next = curr->next;
        free(curr);
        curr = next;
    }
    free(obj);
}

void testLRUCache() {
    printf("=== LRU缓存测试 ===\n");
    
    LRUCache* lru = lRUCacheCreate(2);
    
    lRUCachePut(lru, 1, 1);
    lRUCachePut(lru, 2, 2);
    printf("get(1): %d\n", lRUCacheGet(lru, 1)); // 返回 1
    lRUCachePut(lru, 3, 3); // 该操作会使得关键字 2 作废
    printf("get(2): %d\n", lRUCacheGet(lru, 2)); // 返回 -1 (未找到)
    lRUCachePut(lru, 4, 4); // 该操作会使得关键字 1 作废
    printf("get(1): %d\n", lRUCacheGet(lru, 1)); // 返回 -1 (未找到)
    printf("get(3): %d\n", lRUCacheGet(lru, 3)); // 返回 3
    printf("get(4): %d\n", lRUCacheGet(lru, 4)); // 返回 4
    
    lRUCacheFree(lru);
    
    printf("\n=== C语言特性说明 ===\n");
    printf("1. 手动内存管理: malloc/free\n");
    printf("2. 指针操作: 直接内存访问\n");
    printf("3. 结构体: 数据封装\n");
    printf("4. 哈希表: 自实现的链式哈希\n");
    printf("5. 函数指针: 可扩展设计\n");
}

int main() {
    testLRUCache();
    return 0;
}

/*
C语言实现特点:

1. **内存管理**:
   - 手动malloc/free管理内存
   - 指针操作提供直接内存访问
   - 需要careful处理内存泄漏

2. **数据结构**:
   - 结构体封装数据
   - 双向链表维护顺序
   - 自实现哈希表

3. **性能特点**:
   - 接近硬件的性能
   - 无垃圾回收开销
   - 编译器优化友好

4. **编程特点**:
   - 过程式编程
   - 显式资源管理
   - 类型安全需要程序员保证

算法复杂度:
- get操作: O(1)平均，O(n)最坏
- put操作: O(1)平均，O(n)最坏
- 空间复杂度: O(capacity)
*/